diff -burN gcc-4.3.2/libobjc/Makefile.in gcc-4.3.2-psp/libobjc/Makefile.in
--- gcc-4.3.2/libobjc/Makefile.in
+++ gcc-4.3.2-psp/libobjc/Makefile.in
@@ -72,7 +72,7 @@
 RANLIB = @RANLIB@
 
 CC = @CC@
-CFLAGS = @CFLAGS@
+CFLAGS = -G 0 -G0 @CFLAGS@
 WARN_CFLAGS = -W -Wall -Wwrite-strings -Wstrict-prototypes
 ALL_CFLAGS = -I. -I$(srcdir) $(CPPFLAGS) $(DEFS) $(CFLAGS) $(WARN_CFLAGS) \
        -DIN_GCC -DIN_TARGET_LIBS -fno-strict-aliasing -fexceptions
