diff -Naur gdb-6.8.orig/gdb/tui/tui-io.c gdb-6.8/gdb/tui/tui-io.c
--- gdb-6.8.orig/gdb/tui/tui-io.c	2008-01-01 14:53:22 -0800
+++ gdb-6.8/gdb/tui/tui-io.c	2009-12-01 07:27:14 -0800
@@ -113,8 +113,8 @@
    #undef TUI_USE_PIPE_FOR_READLINE.  */
 
 /* For gdb 5.3, prefer to continue the pipe hack as a backup wheel.  */
-#define TUI_USE_PIPE_FOR_READLINE
-/* #undef TUI_USE_PIPE_FOR_READLINE */
+/* #define TUI_USE_PIPE_FOR_READLINE */
+#undef TUI_USE_PIPE_FOR_READLINE
 
 /* TUI output files.  */
 static struct ui_file *tui_stdout;
