diff -burN newlib-1.15.0/newlib/libc/include/stdint.h
--- newlib-1.15.0/newlib/libc/include/stdint.h	Mon Jun  9 09:00:15 2008
+++ newlib-1.15.0/newlib/libc/include/stdint.h	Mon Jun  9 09:01:04 2008
@@ -34,11 +34,9 @@
 #define __have_longlong64 1
 #endif
 
-/* Check if "long" is 64bit or 32bit wide */
+/* Check if "long" is 64bit */
 #if __STDINT_EXP(LONG_MAX) > 0x7fffffff
 #define __have_long64 1
-#elif __STDINT_EXP(LONG_MAX) == 0x7fffffff && !defined(__SPU__)
-#define __have_long32 1
 #endif
 
 #if __STDINT_EXP(SCHAR_MAX) == 0x7f
