===================================================================
--- psplinkusb/Makefile.oe.orig	2009-06-19 09:06:44.654743748 +0200
+++ psplinkusb/Makefile.oe	2009-06-19 00:30:56.000000000 +0200
@@ -10,7 +10,7 @@
 	$(MAKE) -C boot271   release PSP_FW_VERSION=271
 
 release: clean all
-	mkdir -p release_oe/
+	mkdir -p release_oe
 	mkdir -p release_oe/pc
 	mkdir -p release_oe/psplink
 	cp -Rf scripts release_oe/scriptsIndex: psplinkusb/usbgdb/main.c

