diff -Naur pspsdk.orig/src/base/build.mak pspsdk/src/base/build.mak
--- pspsdk.orig/src/base/build.mak	2010-11-18 14:05:30.947438192 +0100
+++ pspsdk/src/base/build.mak	2010-11-18 09:05:40.657438188 +0100
@@ -17,6 +17,7 @@
 
 CC       = psp-gcc
 CXX      = psp-g++
+GDC      = psp-gdc
 AS       = psp-gcc
 LD       = psp-gcc
 AR       = psp-ar
@@ -46,6 +47,11 @@
 CFLAGS += -D_PSP_FW_VERSION=$(PSP_FW_VERSION)
 CXXFLAGS += -D_PSP_FW_VERSION=$(PSP_FW_VERSION)
 
+# Objective-C selection. All Objective C code must be linked against libobjc.a
+ifeq ($(USE_OBJC),1)
+LIBS     := $(LIBS) -lobjc
+endif
+
 ifeq ($(BUILD_PRX),1)
 LDFLAGS  := $(addprefix -L,$(LIBDIR)) -specs=$(PSPSDK)/lib/prxspecs -Wl,-q,-T$(PSPSDK)/lib/linkfile.prx $(LDFLAGS)
 EXTRA_CLEAN += $(TARGET).elf
@@ -207,6 +213,15 @@
 %.c: %.exp
 	psp-build-exports -b $< > $@
 
+%.o: %.m
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+%.o: %.mm
+	$(CXX) $(CXXFLAGS) -c -o $@ $<
+
+%.o: %.d
+	$(GDC) $(DFLAGS) -c -o $@ $<
+
 clean: 
 	-rm -f $(FINAL_TARGET) $(EXTRA_CLEAN) $(OBJS) $(PSP_EBOOT_SFO) $(PSP_EBOOT) $(EXTRA_TARGETS)
 
diff -Naur pspsdk.orig/src/base/build_prx.mak pspsdk/src/base/build_prx.mak
--- pspsdk.orig/src/base/build_prx.mak	2010-11-18 14:05:30.947438192 +0100
+++ pspsdk/src/base/build_prx.mak	2010-11-18 09:05:40.657438188 +0100
@@ -17,6 +17,7 @@
 
 CC       = psp-gcc
 CXX      = psp-g++
+GDC      = psp-gdc
 AS       = psp-gcc
 LD       = psp-gcc
 FIXUP    = psp-fixup-imports
@@ -37,6 +38,11 @@
 
 CFLAGS += -D_PSP_FW_VERSION=$(PSP_FW_VERSION)
 
+# Objective-C selection. All Objective C code must be linked against libobjc.a
+ifeq ($(USE_OBJC),1)
+LIBS     := $(LIBS) -lobjc
+endif
+
 # Library selection.  By default we link with Newlib's libc.  Allow the
 # user to link with PSPSDK's libc if USE_PSPSDK_LIBC is set to 1.
 
@@ -84,6 +80,15 @@
 %.c: %.exp
 	psp-build-exports -b $< > $@
 
+%.o: %.m
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+%.o: %.mm
+	$(CXX) $(CXXFLAGS) -c -o $@ $<
+
+%.o: %.d
+	$(GDC) $(DFLAGS) -c -o $@ $<
+
 clean: $(EXTRA_CLEAN)
 	-rm -f $(FINAL_TARGET) $(TARGET).elf $(OBJS)
 
diff -Naur pspsdk.orig/src/samples/d/main.d pspsdk/src/samples/d/main.d
--- pspsdk.orig/src/samples/d/main.d	1970-01-01 01:00:00.000000000 +0100
+++ pspsdk/src/samples/d/main.d	2010-11-18 09:05:40.657438188 +0100
@@ -0,0 +1,31 @@
+// imports
+//
+extern (C) void pspDebugScreenInit();
+extern (C) void pspDebugScreenPrintf(char*,...);
+extern (C) int SetupCallbacks();
+extern (C) int sceKernelSleepThread();
+
+import std.string;
+
+class HelloPsp
+{
+public:
+	void sayHello()
+	{
+		pspDebugScreenPrintf(toStringz(m_msg));
+	}
+	char[] m_msg = "Hello D on Psp\n";
+}
+
+int main()
+{
+	SetupCallbacks();
+
+	pspDebugScreenInit();
+	
+	auto hello = new HelloPsp();
+	hello.sayHello();
+
+	sceKernelSleepThread();
+	return 0;
+}
diff -Naur pspsdk.orig/src/samples/d/Makefile.sample pspsdk/src/samples/d/Makefile.sample
--- pspsdk.orig/src/samples/d/Makefile.sample	1970-01-01 01:00:00.000000000 +0100
+++ pspsdk/src/samples/d/Makefile.sample	2010-11-18 09:05:40.657438188 +0100
@@ -0,0 +1,16 @@
+TARGET = HelloPspInD
+OBJS = module.o main.o
+
+CFLAGS = -O2 -Wall
+CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
+ASFLAGS = $(CFLAGS)
+
+LIBDIR =
+LIBS = -lgphobos -lm
+LDFLAGS =
+
+EXTRA_TARGETS = EBOOT.PBP
+PSP_EBOOT_TITLE = Hello World
+
+PSPSDK=$(shell psp-config --pspsdk-path)
+include $(PSPSDK)/lib/build.mak 
diff -Naur pspsdk.orig/src/samples/d/module.c pspsdk/src/samples/d/module.c
--- pspsdk.orig/src/samples/d/module.c	1970-01-01 01:00:00.000000000 +0100
+++ pspsdk/src/samples/d/module.c	2010-11-18 09:05:40.657438188 +0100
@@ -0,0 +1,36 @@
+#include <pspkernel.h>
+#include <pspdebug.h> 
+#include <pspsuspend.h>
+
+PSP_MODULE_INFO("Blit Sample", 0, 1, 1);
+PSP_MAIN_THREAD_ATTR(THREAD_ATTR_USER);
+
+/* Exit callback */
+static int exit_callback(int arg1, int arg2, void *common) {
+	sceKernelExitGame();
+	return 0;
+}
+
+/* Callback thread */
+static int CallbackThread(SceSize args, void *argp) {
+	int cbid;
+
+	cbid = sceKernelCreateCallback("Exit Callback", exit_callback, NULL);
+	sceKernelRegisterExitCallback(cbid);
+
+	sceKernelSleepThreadCB();
+
+	return 0;
+}
+
+/* Sets up the callback thread and returns its thread id */
+int SetupCallbacks(void) {
+	int thid = 0;
+
+	thid = sceKernelCreateThread("update_thread", CallbackThread, 0x11, 0xFA0, 0, 0);
+	if(thid >= 0) {
+		  sceKernelStartThread(thid, 0, 0);
+	}
+
+	return thid;
+} 
diff -Naur pspsdk.orig/src/samples/Makefile.am pspsdk/src/samples/Makefile.am
--- pspsdk.orig/src/samples/Makefile.am	2010-11-18 14:05:31.367438192 +0100
+++ pspsdk/src/samples/Makefile.am	2010-11-18 09:05:40.657438188 +0100
@@ -82,7 +82,9 @@
 	utility/osk \
 	me/basic \
 	wlan \
-	exceptions
+	exceptions \
+	obj-c \
+	d
 
 all:
 
diff -Naur pspsdk.orig/src/samples/obj-c/main.m pspsdk/src/samples/obj-c/main.m
--- pspsdk.orig/src/samples/obj-c/main.m	1970-01-01 01:00:00.000000000 +0100
+++ pspsdk/src/samples/obj-c/main.m	2010-11-18 09:05:40.657438188 +0100
@@ -0,0 +1,62 @@
+#include <objc/Object.h>
+
+@interface Greeter:Object
+{
+ /* This is left empty on purpose:
+  ** Normally instance variables would be declared here,
+  ** but these are not used in our example.
+  */
+}
+
+- (void)greet;
+
+@end
+
+
+#include <pspkernel.h>
+#include <pspdebug.h>
+#include <pspctrl.h>
+
+/* Define printf, just to make typing easier */
+#define printf  pspDebugScreenPrintf
+
+@implementation Greeter
+
+- (void)greet
+{
+	printf("Hello, World from Obj-C!\n");
+}
+
+@end
+
+/* Define the module info section */
+PSP_MODULE_INFO("template", 0, 1, 1);
+
+/* Define the main thread's attribute value (optional) */
+PSP_MAIN_THREAD_ATTR(THREAD_ATTR_USER | THREAD_ATTR_VFPU);
+
+int main(void)
+{
+	/* Non Objective-C code to allow the app to end */
+	pspDebugScreenInit();
+	SceCtrlData pad;
+	
+	/* Objective-C code */
+	id myGreeter;
+	myGreeter=[Greeter new];
+
+	[myGreeter greet];
+	[myGreeter free];
+
+	printf("\nPress X to quit.\n");
+	for (;;)
+	{
+		sceCtrlReadBufferPositive(&pad, 1);
+		if (pad.Buttons & PSP_CTRL_CROSS)
+			break;
+	}
+	sceKernelExitGame();
+	
+	return 0;
+}
+
diff -Naur pspsdk.orig/src/samples/obj-c/Makefile.sample pspsdk/src/samples/obj-c/Makefile.sample
--- pspsdk.orig/src/samples/obj-c/Makefile.sample	1970-01-01 01:00:00.000000000 +0100
+++ pspsdk/src/samples/obj-c/Makefile.sample	2010-11-18 09:05:40.657438188 +0100
@@ -0,0 +1,19 @@
+TARGET = ObjC
+OBJS = main.o
+
+USE_OBJC=1
+
+INCDIR =
+CFLAGS = -G0 -Wall -O2
+CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
+ASFLAGS = $(CFLAGS)
+
+LIBDIR =
+LDFLAGS =
+
+EXTRA_TARGETS = EBOOT.PBP
+PSP_EBOOT_TITLE = ObjC Sample
+
+PSPSDK=$(shell psp-config --pspsdk-path)
+include $(PSPSDK)/lib/build.mak
+
