Index: config/makefile.psp
===================================================================
--- config/makefile.psp	(revision 2366)
+++ config/makefile.psp	(working copy)
@@ -16,7 +16,7 @@
 
 PSPSDK := $(shell psp-config --pspsdk-path)
 
-LINK_OPENGL=-L$(PSPSDK)/lib -lGLU -lGL -lpspdisplay -lpspge -lpsprtc -lpspctrl
+LINK_OPENGL=-L$(PSPSDK)/lib -lGLU -lGL -lpspdisplay -lpspge -lpsprtc -lpspctrl -lpspvfpu
 LINK_MATH= -lstdc++ -lpspdebug -lpspsdk -lm -lc -lpsputility -lpspuser -lpspkernel
 
 # use gcc to link
