Index: libpsp2d/Image.cpp
===================================================================
--- libpsp2d/Image.cpp	(revision 2494)
+++ libpsp2d/Image.cpp	(working copy)
@@ -129,7 +129,7 @@
     if (info_ptr == NULL)
     {
        fclose(fp);
-       png_destroy_read_struct(&png_ptr, png_infopp_NULL, png_infopp_NULL);
+       png_destroy_read_struct(&png_ptr, (png_infopp) NULL, (png_infopp) NULL);
 
        throw ImagePNGException("Could not create PNG info struct");
     }
@@ -137,7 +137,7 @@
     png_init_io(png_ptr, fp);
     png_set_sig_bytes(png_ptr, sig_read);
     png_read_info(png_ptr, info_ptr);
-    png_get_IHDR(png_ptr, info_ptr, &pw, &ph, &bit_depth, &color_type, &interlace_type, int_p_NULL, int_p_NULL);
+    png_get_IHDR(png_ptr, info_ptr, &pw, &ph, &bit_depth, &color_type, &interlace_type, (int*) NULL, (int*) NULL);
 
     _width = (u16)pw;
     _height = (u16)ph;
@@ -152,7 +152,7 @@
        png_set_palette_to_rgb(png_ptr);
 
     if (color_type == PNG_COLOR_TYPE_GRAY && bit_depth < 8)
-       png_set_gray_1_2_4_to_8(png_ptr);
+       png_set_expand_gray_1_2_4_to_8(png_ptr);
 
     if (png_get_valid(png_ptr, info_ptr, PNG_INFO_tRNS))
        png_set_tRNS_to_alpha(png_ptr);
@@ -173,7 +173,7 @@
     if (!_data)
     {
        fclose(fp);
-       png_destroy_read_struct(&png_ptr, png_infopp_NULL, png_infopp_NULL);
+       png_destroy_read_struct(&png_ptr, (png_infopp) NULL, (png_infopp) NULL);
 
        throw ImageException("Memory error");
     }
@@ -192,7 +192,7 @@
     free(rows);
 
     png_read_end(png_ptr, info_ptr);
-    png_destroy_read_struct(&png_ptr, &info_ptr, png_infopp_NULL);
+    png_destroy_read_struct(&png_ptr, &info_ptr, (png_infopp) NULL);
     fclose(fp);
 }
 
