Index: libpsp2d/Screen.cpp
===================================================================
--- libpsp2d/Screen.cpp	(revision 2366)
+++ libpsp2d/Screen.cpp	(working copy)
@@ -226,7 +226,7 @@
 
 u32* Screen::getList()
 {
-    return list;
+    return (u32*) list;
 }
 
 void Screen::accept(DrawableVisitor *v)
@@ -261,7 +261,7 @@
     u16* vram16;
     int bufferwidth;
     int pixelformat;
-    int unknown;
+    //int unknown;
     int i, x, y;
     png_structp png_ptr;
     png_infop info_ptr;
@@ -295,7 +295,7 @@
     png_write_info(png_ptr, info_ptr);
     line = (u8*) malloc(SCREEN_WIDTH * 3);
     sceDisplayWaitVblankStart();  // if framebuf was set with PSP_DISPLAY_SETBUF_NEXTFRAME, wait until it is changed
-    sceDisplayGetFrameBuf((void**)&vram32, &bufferwidth, &pixelformat, &unknown);
+    sceDisplayGetFrameBuf((void**)&vram32, &bufferwidth, &pixelformat, PSP_DISPLAY_SETBUF_IMMEDIATE);
     vram16 = (u16*) vram32;
     for (y = 0; y < SCREEN_HEIGHT; y++) {
        for (i = 0, x = 0; x < SCREEN_WIDTH; x++) {
