Index: src/cal3d/hardwaremodel.cpp
===================================================================
--- src/cal3d/hardwaremodel.cpp	(revision 2432)
+++ src/cal3d/hardwaremodel.cpp	(working copy)
@@ -12,6 +12,7 @@
 #include "config.h"
 #endif
 
+#include <string.h>
 
 #include "cal3d/error.h"
 #include "cal3d/hardwaremodel.h"
Index: src/cal3d/platform.cpp
===================================================================
--- src/cal3d/platform.cpp	(revision 2432)
+++ src/cal3d/platform.cpp	(working copy)
@@ -16,6 +16,8 @@
 // Includes                                                                   //
 //****************************************************************************//
 
+#include <string.h>
+
 #include "cal3d/platform.h"
 
  /*****************************************************************************/
Index: src/cal3d/renderer.cpp
===================================================================
--- src/cal3d/renderer.cpp	(revision 2432)
+++ src/cal3d/renderer.cpp	(working copy)
@@ -16,6 +16,8 @@
 // Includes                                                                   //
 //****************************************************************************//
 
+#include <string.h>
+
 #include "cal3d/error.h"
 #include "cal3d/renderer.h"
 #include "cal3d/coremodel.h"
Index: src/cal3d/submesh.cpp
===================================================================
--- src/cal3d/submesh.cpp	(revision 2432)
+++ src/cal3d/submesh.cpp	(working copy)
@@ -12,6 +12,8 @@
 #include "config.h"
 #endif
 
+#include <string.h>
+
 #include "cal3d/submesh.h"
 #include "cal3d/error.h"
 #include "cal3d/coresubmesh.h"
