Index: src/calc.cpp
===================================================================
--- src/calc.cpp	(revision 2395)
+++ src/calc.cpp	(working copy)
@@ -17,6 +17,7 @@
 #define YYSTYPE float
 #define YYERROR_VERBOSE
 
+#include <string.h>
 #include <cmath>
 #include <cctype>
 
Index: src/tinyxml/tinyxml.cpp
===================================================================
--- src/tinyxml/tinyxml.cpp	(revision 2395)
+++ src/tinyxml/tinyxml.cpp	(working copy)
@@ -20,7 +20,8 @@
 3. This notice may not be removed or altered from any source 
 distribution.
 */
-
+#include <stdlib.h>
+#include <string.h>
 #include "tinyxml.h"
 
 
Index: src/tinyxml/tinyxmlparser.cpp
===================================================================
--- src/tinyxml/tinyxmlparser.cpp	(revision 2395)
+++ src/tinyxml/tinyxmlparser.cpp	(working copy)
@@ -23,6 +23,7 @@
 
 
 #include "tinyxml.h"
+#include <string.h>
 #include <ctype.h>
 
 const char* TiXmlBase::SkipWhiteSpace( const char* p )
Index: src/tinyxml/xmltest.cpp
===================================================================
--- src/tinyxml/xmltest.cpp	(revision 2395)
+++ src/tinyxml/xmltest.cpp	(working copy)
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 #include "tinyxml.h"
 
 //
