Index: libpspirkeyb/Makefile
===================================================================
--- libpspirkeyb/Makefile	(revision 2366)
+++ libpspirkeyb/Makefile	(working copy)
@@ -13,9 +13,10 @@
 include $(PSPSDK)/lib/build.mak
 
 install: all
-	cp libpspirkeyb.a $(PSPSDK)/lib
-	cp include/pspirkeyb.h $(PSPSDK)/include
-	cp include/pspirkeyb_rawkeys.h $(PSPSDK)/include
+	mkdir -p ../../target/psp/sdk/lib ../../target/psp/sdk/include
+	cp libpspirkeyb.a ../../target/psp/sdk/lib
+	cp include/pspirkeyb.h ../../target/psp/sdk/include
+	cp include/pspirkeyb_rawkeys.h ../../target/psp/sdk/include
 
 uninstall:
 	rm -f $(PSPSDK)/lib/libpspirkeyb.a
Index: libpspirkeybprx/Makefile
===================================================================
--- libpspirkeybprx/Makefile	(revision 2366)
+++ libpspirkeybprx/Makefile	(working copy)
@@ -13,8 +13,9 @@
 include $(PSPSDK)/lib/build.mak
 
 install: all
-	cp libpspirkeybprx.a $(PSPSDK)/lib
-	cp include/pspirkeybprx.h $(PSPSDK)/include
+	mkdir -p ../../target/psp/sdk/lib ../../target/psp/sdk/include
+	cp libpspirkeybprx.a ../../target/psp/sdk/lib
+	cp include/pspirkeybprx.h ../../target/psp/sdk/include
 
 uninstall:
 	rm -f $(PSPSDK)/lib/libpspirkeybprx.a
Index: Makefile
===================================================================
--- Makefile	(revision 2452)
+++ Makefile	(working copy)
@@ -32,13 +32,13 @@
 
 install: all
 	$(MAKE) -C libpspirkeyb install
-	mkdir -p $(PSPSDK)/samples/irkeyb
-	mkdir -p $(PSPSDK)/samples/irkeyb/keymap
-	cp libpspirkeyb/keymap/*.ini $(PSPSDK)/samples/irkeyb/keymap
-	cp libpspirkeyb/keymap/README.txt $(PSPSDK)/samples/irkeyb/keymap
-	cp libpspirkeyb/config/pspirkeyb.ini $(PSPSDK)/samples/irkeyb
-	cp asciidemo/main.c $(PSPSDK)/samples/irkeyb
-	cp asciidemo/Makefile.sdk $(PSPSDK)/samples/irkeyb/Makefile
+	mkdir -p ../target/psp/sdk/samples/irkeyb
+	mkdir -p ../target/psp/sdk/samples/irkeyb/keymap
+	cp libpspirkeyb/keymap/*.ini ../target/psp/sdk/samples/irkeyb/keymap
+	cp libpspirkeyb/keymap/README.txt ../target/psp/sdk/samples/irkeyb/keymap
+	cp libpspirkeyb/config/pspirkeyb.ini ../target/psp/sdk/samples/irkeyb
+	cp asciidemo/main.c ../target/psp/sdk/samples/irkeyb
+	cp asciidemo/Makefile.sdk ../target/psp/sdk/samples/irkeyb/Makefile
 
 uninstall:
 	$(MAKE) -C libpspirkeyb install
Index: pspirkeybprx/Makefile
===================================================================
--- pspirkeybprx/Makefile	(revision 2366)
+++ pspirkeybprx/Makefile	(working copy)
@@ -23,9 +23,10 @@
 include $(PSPSDK)/lib/build.mak
 
 install: all
-	cp libpspirkeyb.a $(PSPSDK)/lib
-	cp include/pspirkeyb.h $(PSPSDK)/include
-	cp include/pspirkeyb_rawkeys.h $(PSPSDK)/include
+	mkdir -p ../../target/psp/sdk/lib ../../target/psp/sdk/include
+	cp libpspirkeyb.a ../../target/psp/sdk/lib
+	cp include/pspirkeyb.h ../../target/psp/sdk/include
+	cp include/pspirkeyb_rawkeys.h ../../target/psp/sdk/include
 
 uninstall:
 	rm -f $(PSPSDK)/lib/libpspirkeyb.a
Index: asciidemo/main.c
===================================================================
--- asciidemo/main.c	(revision 2452)
+++ asciidemo/main.c	(working copy)
@@ -9,7 +9,6 @@
 #include <pspkernel.h>
 #include <pspdebug.h>
 #include <pspctrl.h>
-#include <kubridge.h>
 
 #include <pspirkeyb.h>
 #include <pspirkeyb_rawkeys.h>
Index: asciidemo/Makefile
===================================================================
--- asciidemo/Makefile	(revision 2452)
+++ asciidemo/Makefile	(working copy)
@@ -9,7 +9,7 @@
 LIBDIR = $(PSPDEV)/SDK/lib ../libpspirkeyb
 LDFLAGS =
 
-LIBS = -lpspkubridge -lpspirkeyb -lpsppower
+LIBS = -lpspirkeyb -lpsppower
 
 BUILD_PRX = 1
 PSP_FW_VERSION = 401
Index: asciidemoprx/main.c
===================================================================
--- asciidemoprx/main.c	(revision 2452)
+++ asciidemoprx/main.c	(working copy)
@@ -13,7 +13,6 @@
 #include <pspctrl.h>
 #include <pspdebug.h>
 #include <psppower.h>
-#include <kubridge.h>
 
 #include <stdio.h>
 #include <string.h>
Index: asciidemoprx/Makefile
===================================================================
--- asciidemoprx/Makefile	(revision 2452)
+++ asciidemoprx/Makefile	(working copy)
@@ -9,7 +9,7 @@
 LIBDIR = $(PSPDEV)/SDK/lib ../libpspirkeybprx
 LDFLAGS =
 
-LIBS = -lpspkubridge -lpspirkeybprx -lpsppower
+LIBS = -lpspirkeybprx -lpsppower
 
 BUILD_PRX = 1
 PSP_FW_VERSION = 401
